azurelinuxagent/common/protocol/extensions_config_retriever.py

```diff
@@ -0,0 +1,374 @@
+# Microsoft Azure Linux Agent
+#
+# Copyright Microsoft Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Requires Python 2.6+ and Openssl 1.0+
+#
+
+import os
+import re
+
+from azurelinuxagent.common.future import ustr
+
+import azurelinuxagent.common.conf as conf
+import azurelinuxagent.common.logger as logger
+from azurelinuxagent.common.protocol.goal_state import ExtensionsConfig
+from azurelinuxagent.common.utils import fileutil, restutil
+from azurelinuxagent.common.exception import ProtocolError
+from azurelinuxagent.common.utils.shellutil import run_command
+from azurelinuxagent.common.utils.textutil import safe_shlex_split
+
+INCARNATION_FILE_NAME = "Incarnation"
+SEQUENCE_NUMBER_FILE_NAME = "ArtifactProfileSequenceNumber"
+SVD_SEQNO_FILE_NAME = "SvdSeqNo"
+GOAL_STATE_SOURCE_FILE_NAME = "GoalStateSource"
+VM_ID_FILE_NAME = "VmId"
+
+GOAL_STATE_SOURCE_FABRIC = "Fabric"
+GOAL_STATE_SOURCE_FASTTRACK = "FastTrack"
+
+DMIDECODE_CALL = "dmidecode"
+
+
+class ExtensionsConfigReasons:
+    FABRIC_CHANGED = "FabricChanged"
+    FABRIC_LAST_CHANGE = "LastFabric"
+    FAST_TRACK_CHANGED = "FTChanged"
+    FAST_TRACK_LAST_CHANGE = "LastFT"
+
+
+class FastTrackChangeDetail:
+    NO_CHANGE = "NoChange"
+    NO_EXTENSIONS = "NoExtensions"
+    NO_PROFILE = "NoProfile"
+    SEQ_NO_CHANGED = "seqNoChanged"
+
+
+class FabricChangeDetail:
+    INCARNATION_CHANGED = "IncChanged"
+    NO_CHANGE = "NoChange"
+    NO_INCARNATION = "NoInc"
+
+
+""" 
+GenericExtensionsConfig abstracts whether we pulled the goal state from Fabric or from FastTrack
+consumers should not worry from where the ExtensionsConfig came. They should also have no knowledge
+of sequence numbers or incarnations, which are specific to FastTrack and Fabric respectfully
+"""
+class GenericExtensionsConfig(ExtensionsConfig):
+    def __init__(self, extensions_config, changed, ext_conf_retriever):
+        self.extensions_config = extensions_config
+        self.changed = changed
+        self._ext_conf_retriever = ext_conf_retriever
+
+        ExtensionsConfig.__init__(self, extensions_config.xml_text)
```

Comment:
This line is not needed. You dont have to call the parent constructor in python, it works without it too.

Eg: 
```python
class A(object):
...     def __init__(self, prop=123):
...         self.prop = prop
... 
class B(A):
...     def __init__(self, prop=456):
...         self.prop = prop
... 
obj = B()
obj.prop
456
obj = A()
obj.prop
123
```

Reply:
In this case it's required because I'm passing extensions_config.xml_text to the base class

Reply:
Why is that needed though? The `ExtensionsConfig.__init__` function mainly parses the xml and returns an object but in this case we're not using it anywhere.

Reply:
We treat GenericExtensionsConfig as an ExtensionsConfig everywhere. In fact, if I could make ExtensionsConfig abstract (possible in Python?) I would. Therefore we need GenericExtensionsConfig to have all of the same properties.